docs: AD Site Discovery

justin-stephenson opened the pull request:
The ``Site-enabled discovery`` section numbered list is broken up causing the numbers to be incorrect, can someone help with how to fix this

For the reviewer:

* Should the ``Non-site discovery`` section be included or not? Is it helpful?
* I did not add mention trusted domains at all, is this relevant for this document, or another document perhaps ?

pbrezina commented:
This works for me. You need to be pedantic on the indentation there.

```diff
diff --git a/src/docs/ad/ad-site-discovery.rst b/src/docs/ad/ad-site-discovery.rst
index 812fa9d..875578d 100644
--- a/src/docs/ad/ad-site-discovery.rst
+++ b/src/docs/ad/ad-site-discovery.rst
@@ -6,46 +6,47 @@ Site-enabled discovery
 ======================
 At domain initialization, if ``ad_enable_dns_sites`` is true (default) then the 'AD' SRV lookup plugin is set. SSSD will attempt to discover primary(site-specific) servers and backup servers using the following steps:
 
-#.  **Set discovery domain**: read from ``dns_discovery_domain`` option or use ``ad_domain`` name
-
+#. **Set discovery domain**: read from ``dns_discovery_domain`` option or use ``ad_domain`` name
 #. **Lookup AD site and forest** information using CLDAP ping, or read ``ad_site`` option in ``sssd.conf``
 
-  .. note::
+   * Find DCs to perform CLDAP ping against. This is done by :ref:`dns-srv-resolution`. using the domain from step 1 and 'ldap' service as inputs.
+   * CLDAP ping is executed in 3 "batches" to avoid pinging too many domain controllers and avoid timeouts when attempting to ping unreachable AD DCs. If a response is received at any step then we skip the remaining steps:
 
-     C(Connection-Less)LDAP Ping is a way for LDAP clients to request server information, it is used by Microsoft Clients against AD.
+     * Batch 1: CLDAP ping up to the first 5 domain controllers in the resolved server list
+     * Wait 400ms for a response
+     * Batch 2: CLDAP ping up to the next 5 domain controllers
+     * Wait 200ms for a response
+     * Batch 3: CLDAP Ping any remaining domain controllers
 
-  .. note::
+#. **Discover primary and backup servers**
 
-     SSSD tries to reuse previously stored AD site and forest information, CLDAP pings are only performed if SSSD is starting up, or recovering from an offline state
+   If a site was found above, discover **site-specific** servers to use as primary servers, backup servers will be any servers discovered in the domain (not in the site)
 
-  * Find DCs to perform CLDAP ping against. This is done by :ref:`dns-srv-resolution`. using the domain from step 1 and 'ldap' service as inputs.
-  * CLDAP ping is executed in 3 "batches" to avoid pinging too many domain controllers and avoid timeouts when attempting to ping unreachable AD DCs. If a response is received at any step then we skip the remaining steps:
+   * Primary servers will be the result of **site-specific** :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
+   * Backup servers will be the result of :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
 
-       * Batch 1: CLDAP ping up to the first 5 domain controllers in the resolved server list
-       * Wait 400ms for a response
-       * Batch 2: CLDAP ping up to the next 5 domain controllers
-       * Wait 200ms for a response
-       * Batch 3: CLDAP Ping any remaining domain controllers
+   If no site was found above
 
-#. **Discover primary and backup servers**
-    If a site was found above, discover **site-specific** servers to use as primary servers, backup servers will be any servers discovered in the domain (not in the site)
+   * Primary servers will be the result of :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
 
-    * Primary servers will be the result of **site-specific** :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
-    * Backup servers will be the result of :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
+#. **Sort Resulting Server list by DNS**
 
-    If no site was found above
+   * Prioritize servers with the same domain as the client, when priority is shared among server list result
 
-    * Primary servers will be the result of :ref:`dns-srv-resolution` for dns discovery domain and 'ldap' service
+#. Once this sequence of operations are done, SSSD saves
 
-#. **Sort Resulting Server list by DNS**
-        * Prioritize servers with the same domain as the client, when priority is shared among server list result
+   * a sorted list of primary servers
+   * a sorted list of backup servers, if applicable
+   * a resolved DNS discovery domain
+   * a DNS reply TTL
 
-#. Once this sequence of operations are done, SSSD saves
+.. note::
+
+    C(Connection-Less)LDAP Ping is a way for LDAP clients to request server information, it is used by Microsoft Clients against AD.
+
+    .. note::
 
-        * a sorted list of primary servers
-        * a sorted list of backup servers, if applicable
-        * a resolved DNS discovery domain
-        * a DNS reply TTL
+    SSSD tries to reuse previously stored AD site and forest information, CLDAP pings are only performed if SSSD is starting up, or recovering from an offline state
 
 
 Non-site discovery

```

pbrezina commented:
Given the document has little wider range then just site discovery I suggest:

* Rename to ad-service-discovery.rst
* Title "Active Directory Service Discovery", in navigation shorten to "Service Discovery"
* Remove DNS SRV resolution and timeouts - this is quite generic mechanism and it would be good to describe it together with whole failover in new document (please add it to the planning sheet)
* Remove "non site discovery" this is not widely used (if ever) and in general not recommended

Make the article more descriptive so the reader can get basic understanding of what a DC, GC, site and forest (which is also discovered in this step) is (imaging new SSSD developer or tester) and back it up with links to Microsoft documentation for a thorough explanation.

CLDAP is LDAP over UDP, CLDAP or LDAP ping checks if DC is alive and in addition we request specific netlogon information (see https://ldapwiki.com/wiki/LDAP%20ping for more information), I think more information about it is helpful. You can also link a Microsoft specification to provide information how it is done on Windows, see: https://github.com/SSSD/sssd/pull/5300#issuecomment-700577606

Backing the algorithm with a diagram would be also nice (probably in Lucid charts since it gives you more options, feel free to ping me if you need any help).

You can also add information about manual setting of ad site (ad_site) and that the site is cache to speed up the next start.

justin-stephenson commented:
Changes made as requested, please see updated doc.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+CLDAP Ping
+    See section below
+
+When using the AD provider, by default SSSD clients utilize auto-discovery of Active Directory servers to help avoid explicitly configuring Active Directory servers, and enable failover behavior. The term **Service Discovery** may also be used in SSSD documentation and manual pages to describe this.
```
"Active Directory servers to help avoid explicitly configuring Active Directory servers" -> "Active Directory servers to avoid the need to explicitly define them on sssd.conf on each client."?

Also explain the goal? Ie to discover global catalog and domain controlles that will be used as primary and backup servers. Little explanation of primary servers (in-site) and backup (off-site when primary are not available)?

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+
+LDAP/CLDAP Ping
+===============
+`C(Connection-Less)LDAP Ping <https://ldapwiki.com/wiki/LDAP%20ping>`_ is a way for clients to check the liveliness of a domain controller. It is often necessary for SSSD to try to find available domain controllers within a domain, or within an AD Site. SSSD sends a LDAP Ping during the service discovery phase by sending a `rootDSE <https://ldapwiki.com/wiki/RootDSE>`_ search of the **netlogon**  ...
```
CLDAP (Connection-less) ping is a way for clients to check the liveliness of a domain controller. Instead of LDAP where the communication is done over TCP protocol, the CLDAP ping is sent over UDP to be faster. (ie mention it is over UDP not TCP)

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+
+    .. note::
+
+    SSSD tries to reuse previously stored AD site and forest information, CLDAP pings are only performed if SSSD is starting up, or recovering from an offline state
```
I didn't see the code for a long time, but I think we cache it this way (please check):

* Store found site in cache
* If cached, prefer in-site dcs in CLDAP ping

justin-stephenson replied:
Correct, I updated the note to describe this behavior better.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+
+Site-enabled discovery
+======================
+At domain initialization, if ``ad_enable_dns_sites`` is true (default) then the 'AD' SRV lookup plugin is set. SSSD will attempt to discover primary(site-specific) servers and backup servers using the following steps:
```
\``AD``

primary <space> (in-site)

justin-stephenson replied:
Should be fixed now.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. note::
+
+    The **ad_site** option can be set manually in the sssd configuration to speed up the discovery process.
```
\``ad_site``

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+
+.. image:: cldap-ping.svg
+   :height: 700px
+   :align: center
```
Nice image! But it is distorted, you need to set smaller height.

Make the Batch 3 arrow connect with batch 2 and 1 arrows?

pbrezina reviewed (changes requested):
Thank you. Some comments are inline, but you did a good job!

justin-stephenson commented:
> Thank you. Some comments are inline, but you did a good job!

Thanks very much for the review, please see latest updates.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,84 @@
+`Active Directory <https://en.wikipedia.org/wiki/Active_Directory>`_ is a directory service providing Windows and Linux clients with information about centrally stored user and group information. Below are some Active Directory terms that are important to understand for the purposes of this document.
+
+Domain Controller
+    The Active Directory Domain Controller (DC for short) is an important source of user information and authentication in many computing environments. The DC stores user and group information in a LDAP directory server (database) and provides this information as a `service <https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/get-started/virtual-dc/active-directory-domain-services-over ...
```
a LDAP -> an LDAP?

justin-stephenson replied:
I don't believe so - https://owl.purdue.edu/owl/general_writing/grammar/articles_a_versus_an.html

pbrezina replied:
Isn't it driven by the sound instead of written form? I.e. "an eldap" instead of "a eldap", like "an honorable piece" or "a union" from the listed exceptions?

pbrezina replied:
I don't want to teach you your language, this is what I have been taught so I'm curios.

justin-stephenson replied:
You are right, "If the first letter makes a vowel-type sound" I was just looking at LDAP but not thinking about this. Fixed to 'an LDAP'

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,84 @@
+    A logical organization, or container for objects within a domain. `Sites <https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/plan/understanding-active-directory-site-topology>`_ are often used to group geographically-close systems together. Clients can prefer to communicate only to servers within a site.
+
+Global Catalog
+    A service advertising information about Active Directory domains in a forest, it can be searched by clients similarly to normal LDAP user/group information. The `Global Catalog <https://docs.microsoft.com/en-us/windows/win32/ad/global-catalog>`_ is used by SSSD clients to read AD Site and Forest information, and trusted domains.
```
The netlogon attribute that is used to get site and forest names is present on every DC and we look it up in DC not GC. Some domain controllers are global catalogs, not all domain controllers are global catalogs. Global catalog is a little bit more difficult thing, we use it to lookup user and group information (not the site and forest). Perhaps it is not needed to mention it here and complicate stuff.

pbrezina replied:
It does a lot more then advertising information about domains which is not needed for the sake of this article, so I'd vote to remove it. In short, global catalog contains information (selected attributes) of every object in the forest (from all domains).

justin-stephenson replied:
Okay makes sense, I removed 'global catalog' from the terminology list.

pbrezina reviewed (changes requested):
Thanks for the update. Some stuff were not addressed so I kept them open and added two minor things.

Additionally, please make sure:
* be consistent with SSSD letter size (preferably all capitals) - don't mix SSSD and sssd
* space belongs before opening parentheses unless i.e. primary (in-site) not primary(in-site). The latter denotes function call.
* Site when in the middle of a sentence is lowercase "site"
* I think also CLDAP Ping sohuld be CLDAP ping

justin-stephenson commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,87 @@
+    A logical organization, or container for objects within a domain. `Sites <https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/plan/understanding-active-directory-site-topology>`_ are often used to group geographically-close systems to together. Clients can prefer to communicate only to servers within a site.
+
+Global Catalog
+    A service advertising information about Active Directory domains in a forest, it can be searched by clients similarly to normal LDAP user/group information. The `Global catalog <https://docs.microsoft.com/en-us/windows/win32/ad/global-catalog>`_ is used by SSSD clients to read and discovery AD Site and Forest information, and trusted domains.
```
Removed this section.

justin-stephenson commented:
> Thanks for the update. Some stuff were not addressed so I kept them open and added two minor things.
> 
> Additionally, please make sure:
> 
>     * be consistent with SSSD letter size (preferably all capitals) - don't mix SSSD and sssd

Fixed.
> 
>     * space belongs before opening parentheses unless i.e. primary (in-site) not primary(in-site). The latter denotes function call.

Fixed, bad habit I'm trying to work on :)
> 
>     * Site when in the middle of a sentence is lowercase "site"

Fixed.
> 
>     * I think also CLDAP Ping sohuld be CLDAP ping

Fixed.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,81 @@
+
+    C(Connection-Less)LDAP ping behavior performed by Microsoft clients is described in `Microsoft protocol documentation <https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/WinArchive/%5bMS-DISO%5d.pdf>`_
+
+    .. note::
```
Deindent/removed based on what your intention is.

justin-stephenson replied:
Fixed.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,81 @@
+
+LDAP/CLDAP ping
+===============
+`C(Connection-Less)LDAP ping <https://ldapwiki.com/wiki/LDAP%20ping>`_ is a way for clients to check the liveliness of a domain controller. Instead of LDAP where the communication is done over TCP protocol, the CLDAP ping is sent over UDP to be faster. It is often necessary for SSSD to try to find available domain controllers within a domain, or within an AD site. SSSD sends a LDAP ping during th ...
```
I'd prefer CLDAP (Connection-less LDAP) but it is up to you.

justin-stephenson replied:
Fixed.

pbrezina commented on src/docs/ad/ad-service-discovery.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. note::
+
+    C(Connection-Less)LDAP ping behavior performed by Microsoft clients is described in `Microsoft protocol documentation <https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/WinArchive/%5bMS-DISO%5d.pdf>`_
```
I think CLDAP is enough here since you already explained what C stands for in the paragraph above.

justin-stephenson replied:
Fixed.

pbrezina reviewed (changes requested):
This is a final review, lets decide on the article before LDAP and last comments and we can push.

pbrezina commented:
Good job. Ack.

justin-stephenson commented:
@sumit-bose Peter asked me to make you aware of this document, in case it would be useful for your sssd AD training.
